Mathlib/AlgebraicGeometry/Cover/MorphismProperty.lean

```diff
@@ -0,0 +1,342 @@
+    let y := (𝒰.covers x).choose
+    have hy : (𝒰.map (𝒰.f x)).base y = x := (𝒰.covers x).choose_spec
+    rcases (f (𝒰.f x)).covers y with ⟨z, hz⟩
+    change x ∈ Set.range ((f (𝒰.f x)).map ((f (𝒰.f x)).f y) ≫ 𝒰.map (𝒰.f x)).base
+    use z
+    erw [comp_apply]
+    rw [hz, hy]
+  map_prop _ := P.comp_mem _ _ ((f _).map_prop _) (𝒰.map_prop _)
+
+/-- An isomorphism `X ⟶ Y` is a `P`-cover of `Y`. -/
+@[simps J obj map]
+def coverOfIsIso [P.ContainsIdentities] [P.RespectsIso] {X Y : Scheme.{u}} (f : X ⟶ Y)
+    [IsIso f] : Cover.{v} P Y where
+  J := PUnit.{v + 1}
+  obj _ := X
+  map _ := f
+  f _ := PUnit.unit
+  covers x := by
+    rw [Set.range_iff_surjective.mpr]
+    all_goals try trivial
+    rw [← TopCat.epi_iff_surjective]
+    infer_instance
+  map_prop _ := P.of_isIso _
+
+/-- We construct a cover from another, by providing the needed fields and showing that the
+provided fields are isomorphic with the original cover. -/
+@[simps J obj map]
+def Cover.copy [P.RespectsIso] {X : Scheme.{u}} (𝒰 : X.Cover P)
+    (J : Type*) (obj : J → Scheme)
+    (map : ∀ i, obj i ⟶ X) (e₁ : J ≃ 𝒰.J) (e₂ : ∀ i, obj i ≅ 𝒰.obj (e₁ i))
+    (h : ∀ i, map i = (e₂ i).hom ≫ 𝒰.map (e₁ i)) : X.Cover P :=
+  { J, obj, map
+    f := fun x ↦ e₁.symm (𝒰.f x)
+    covers := fun x ↦ by
+      rw [h, Scheme.comp_base, TopCat.coe_comp, Set.range_comp, Set.range_iff_surjective.mpr,
+        Set.image_univ, e₁.rightInverse_symm]
+      · exact 𝒰.covers x
+      · rw [← TopCat.epi_iff_surjective]; infer_instance
+    map_prop := fun j ↦ by
+      rw [h, P.cancel_left_of_respectsIso]
+      exact 𝒰.map_prop (e₁ j) }
+
+/-- The pushforward of a cover along an isomorphism. -/
+@[simps! J obj map]
+def Cover.pushforwardIso [P.RespectsIso] [P.ContainsIdentities] [P.IsStableUnderComposition]
+    {X Y : Scheme.{u}} (𝒰 : Cover.{v} P X) (f : X ⟶ Y) [IsIso f] :
+    Cover.{v} P Y :=
+  ((coverOfIsIso.{v, u} f).bind fun _ => 𝒰).copy 𝒰.J _ _
+    ((Equiv.punitProd _).symm.trans (Equiv.sigmaEquivProd PUnit 𝒰.J).symm) (fun _ => Iso.refl _)
+    fun _ => (Category.id_comp _).symm
+
+/-- Adding map satisfying `P` into a cover gives another cover. -/
+@[simps]
+def Cover.add {X Y : Scheme.{u}} (𝒰 : X.Cover P) (f : Y ⟶ X) (hf : P f := by infer_instance) :
+    X.Cover P where
+  J := Option 𝒰.J
+  obj i := Option.rec Y 𝒰.obj i
+  map i := Option.rec f 𝒰.map i
+  f x := some (𝒰.f x)
+  covers := 𝒰.covers
+  map_prop j := by
+    obtain ⟨_ | _⟩ := j
+    · exact hf
+    · exact 𝒰.map_prop _
+
+/-- A morphism property of schemes is said to preserve joint surjectivity, if
+for any pair of morphisms `f : X ⟶ S` and `g : Y ⟶ S` where `g` satisfies `P`,
+any pair of points `x : X` and `y : Y` with `f x = g y` can be lifted to a point
+of `X ×[S] Y`.
+
+In later files, this will be automatic, since this holds for any morphism `g`. But at
+this early stage in the import tree, we only know it for open immersions. -/
+class IsJointlySurjectivePreserving (P : MorphismProperty Scheme.{u}) where
```

Comment:
Is this instance provided for all morphism properties?

Reply:
I forgot that part, thanks! Fixed.

Reply:
Can you point to this file in the docstring of `IsJointlySurjectivePreserving`?

Reply:
Done